sw/airborne/modules/ins/ins_lpe.c

```diff
@@ -0,0 +1,533 @@
+    float_mat_zero(ins_lpe.B.matrix, ins_lpe.B.rows, ins_lpe.B.cols); // initialize matrix to zero
+    ins_lpe.B.initialized = TRUE; // matrix is ready to be used
+
+    // initialize measurement noise covariance matrix R 3x3
+    ins_lpe.R.initialized = FALSE; // not initialized yet
+    ins_lpe.R.rows = n_u; // set n_rows
+    ins_lpe.R.cols = n_u; // set n_cols
+    POPULATE_MATRIX_PTR(ins_lpe.R_data_ptr, ins_lpe.R_data, ins_lpe.R.rows); // populate the matrix row pointer
+    ins_lpe.R.matrix = &ins_lpe.R_data_ptr[0]; // set the matrix pointer
+    float_mat_zero(ins_lpe.R.matrix, ins_lpe.R.rows, ins_lpe.R.cols); // initialize matrix to zero
+    ins_lpe.R.initialized = TRUE; // matrix is ready to be used
+
+    // initialize process noise covariance Q 9x9
+    ins_lpe.Q.initialized = FALSE; // not initialized yet
+    ins_lpe.Q.rows = n_x; // set n_rows
+    ins_lpe.Q.cols = n_x; // set n_cols
+    POPULATE_MATRIX_PTR(ins_lpe.Q_data_ptr, ins_lpe.Q_data, ins_lpe.Q.rows); // populate the matrix row pointer
+    ins_lpe.Q.matrix = &ins_lpe.Q_data_ptr[0]; // set the matrix pointer
+    float_mat_zero(ins_lpe.Q.matrix, ins_lpe.Q.rows, ins_lpe.Q.cols); // initialize matrix to zero
+    ins_lpe.Q.initialized = TRUE; // matrix is ready to be used
+
+    // initialize estimate error covariance P 9x9
+    ins_lpe.P.initialized = FALSE; // not initialized yet
+    ins_lpe.P.rows = n_x; // set n_rows
+    ins_lpe.P.cols = n_x; // set n_cols
+    POPULATE_MATRIX_PTR(ins_lpe.P_data_ptr, ins_lpe.P_data, ins_lpe.P.rows); // populate the matrix row pointer
+    ins_lpe.P.matrix = &ins_lpe.P_data_ptr[0]; // set the matrix pointer
+    float_mat_zero(ins_lpe.P.matrix, ins_lpe.P.rows, ins_lpe.P.cols); // initialize matrix to zero
+    ins_lpe.P.initialized = TRUE; // matrix is ready to be used
+
+    // initialize state vector 9x1
+    memset(ins_lpe.x, 0, sizeof(ins_lpe.x));
+
+    // initialize input vector 3x1
+    memset(ins_lpe.u, 0, sizeof(ins_lpe.u));
+}
+
+
+void ins_lpe_cov_propagation_logic(float **dP)
+{
+  for (int i = 0; i < n_x; i++) {
+    if (ins_lpe.P.matrix[i][i] > P_MAX) {
+      // if diagonal element greater than max, stop propagating
+      dP[i][i] = 0;
+
+      for (int j = 0; j < n_x; j++) {
+        dP[i][j] = 0;
+        dP[j][i] = 0;
+      }
+    }
+  }
+}
+
+
+
+/**
+ * Prediction step in the Kalman filter
+ */
+void ins_lpe_predict(void)
+{
+  if (ins_lpe.integrate) {
+    // accel in NED
+    ins_lpe.u[U_ax] = ins_lpe.accel.accel_measured.x;
+    ins_lpe.u[U_ay] = ins_lpe.accel.accel_measured.y;
+    ins_lpe.u[U_az] = ins_lpe.accel.accel_measured.z + 9.81f; // add g
+  }
+  else {
+    // zero input
+    ins_lpe.u[U_ax] = 0;
+    ins_lpe.u[U_ay] = 0;
+    ins_lpe.u[U_az] = 0;
+  }
+
+  // update A with new body-to-NED roration matrix
+  ins_lpe_update_ssstates();
+
+
+  // continuous time kalman filter prediction
+  // integrate runge kutta 4th order
+  // TODO move rk4 algorithm to matrixlib
```

Comment:
there is already a RK lib in paparazzi: https://github.com/paparazzi/paparazzi/blob/master/sw/airborne/math/pprz_rk_float.h
it is used in the ins_invariant filter: https://github.com/paparazzi/paparazzi/blob/master/sw/airborne/subsystems/ins/ins_float_invariant.c#L347